File: sdks/typescript/src/ipc/hello.test.ts
import { describe, expect, test } from "bun:test";
import { encodeHello, HELLO_MESSAGE, type HelloRefusalReason, parseHello } from "./hello.js";

describe("encodeHello", () => {
  test("emits the canonical frame, with no trailing newline", () => {
    // The LF is the framing layer's, not this function's — see spec/wire/v1/framing.md §3.
    expect(encodeHello(["1"])).toBe('{"nimbus":"hello","contractVersions":["1"]}');
  });

  test("round-trips through parseHello", () => {
    const parsed = parseHello(encodeHello(["1", "2"]));
    expect(parsed).toEqual({ ok: true, contractVersions: ["1", "2"] });
  });

  test("the discriminator is a published constant", () => {
    expect(HELLO_MESSAGE).toBe("hello");
  });
});

describe("parseHello — the frame is JSON, not a byte pattern", () => {
  test("accepts the canonical form", () => {
    expect(parseHello('{"nimbus":"hello","contractVersions":["1"]}')).toEqual({
      ok: true,
      contractVersions: ["1"],
    });
  });

  test("accepts insignificant whitespace", () => {
    expect(parseHello('{"nimbus": "hello", "contractVersions": ["1"]}')).toEqual({
      ok: true,
      contractVersions: ["1"],
    });
  });

  test("accepts reversed member order", () => {
    expect(parseHello('{"contractVersions":["1"],"nimbus":"hello"}')).toEqual({
      ok: true,
      contractVersions: ["1"],
    });
  });

  test("ignores unknown members", () => {
    expect(parseHello('{"nimbus":"hello","contractVersions":["1"],"extra":{"a":1}}')).toEqual({
      ok: true,
      contractVersions: ["1"],
    });
  });
});

describe("parseHello — refusals", () => {
  const cases: ReadonlyArray<readonly [string, HelloRefusalReason]> = [
    ["{oops", "not-json"],
    ["", "not-json"],
    ["null", "not-object"],
    ['["1"]', "not-object"],
    ["42", "not-object"],
    ['{"nimbus":"goodbye","contractVersions":["1"]}', "wrong-message"],
    ['{"contractVersions":["1"]}', "wrong-message"],
    ['{"nimbus":"hello"}', "missing-versions"],
    ['{"nimbus":"hello","contractVersions":"1"}', "missing-versions"],
    ['{"nimbus":"hello","contractVersions":[]}', "empty-versions"],
    ['{"nimbus":"hello","contractVersions":["01"]}', "invalid-version"],
    ['{"nimbus":"hello","contractVersions":["1.0"]}', "invalid-version"],
    ['{"nimbus":"hello","contractVersions":["\\u0661"]}', "invalid-version"],
    ['{"nimbus":"hello","contractVersions":[1]}', "invalid-version"],
    ['{"nimbus":"hello","contractVersions":["1","1"]}', "duplicate-version"],
  ];

  for (const [frame, reason] of cases) {
    test(`${JSON.stringify(frame)} → ${reason}`, () => {
      expect(parseHello(frame)).toEqual({ ok: false, reason });
    });
  }

  test("never throws, whatever the frame contains", () => {
    for (const frame of ["", " ", "{", "}", '{"nimbus":', "\uD800"]) {
      expect(() => parseHello(frame)).not.toThrow();
    }
  });
});

File: sdks/go/ipc/hello_test.go
package ipc

import "testing"

func TestEncodeHelloProducesTheCanonicalFrame(t *testing.T) {
	got := EncodeHello([]string{"1"})
	want := `{"nimbus":"hello","contractVersions":["1"]}`
	if got != want {
		t.Errorf("got %s, want %s", got, want)
	}
}

func TestParseHelloAcceptsTheCanonicalFrame(t *testing.T) {
	got := ParseHello(`{"nimbus":"hello","contractVersions":["1","2"]}`)
	ok, isOk := got.(HelloOk)
	if !isOk {
		t.Fatalf("got %#v, want HelloOk", got)
	}
	// The frame's declared order is what a parser reports, even though §4 makes the
	// set unordered for negotiation.
	if len(ok.ContractVersions) != 2 || ok.ContractVersions[0] != "1" || ok.ContractVersions[1] != "2" {
		t.Errorf("ContractVersions = %#v, want [1 2]", ok.ContractVersions)
	}
}

func TestParseHelloIgnoresWhitespaceAndMemberOrder(t *testing.T) {
	got := ParseHello("  { \"contractVersions\" : [ \"1\" ] , \"nimbus\" : \"hello\" }  ")
	if _, isOk := got.(HelloOk); !isOk {
		t.Errorf("got %#v, want HelloOk — this parses JSON, not bytes", got)
	}
}

func TestParseHelloRefusalReasons(t *testing.T) {
	tests := []struct {
		name, frame, reason string
	}{
		{"not json", "{", "not-json"},
		{"array", `["1"]`, "not-object"},
		{"null", "null", "not-object"},
		{"wrong discriminator", `{"nimbus":"goodbye","contractVersions":["1"]}`, "wrong-message"},
		{"versions absent", `{"nimbus":"hello"}`, "missing-versions"},
		{"versions not an array", `{"nimbus":"hello","contractVersions":"1"}`, "missing-versions"},
		{"versions empty", `{"nimbus":"hello","contractVersions":[]}`, "empty-versions"},
		{"member not a version", `{"nimbus":"hello","contractVersions":["01"]}`, "invalid-version"},
		{"member not a string", `{"nimbus":"hello","contractVersions":[1]}`, "invalid-version"},
		{"duplicate", `{"nimbus":"hello","contractVersions":["1","1"]}`, "duplicate-version"},
		// Validity is checked per member BEFORE duplication, so this is
		// invalid-version and not duplicate-version.
		{"invalid duplicated", `{"nimbus":"hello","contractVersions":["01","01"]}`, "invalid-version"},
		// Go's encoding/json refuses these already; Python needs a hook to match.
		{"NaN", `{"nimbus":"hello","contractVersions":[NaN]}`, "not-json"},
		{"Infinity", `{"nimbus":"hello","contractVersions":[Infinity]}`, "not-json"},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := ParseHello(tt.frame)
			refused, isRefused := got.(HelloRefused)
			if !isRefused {
				t.Fatalf("got %#v, want HelloRefused{%s}", got, tt.reason)
			}
			if refused.Reason != tt.reason {
				t.Errorf("Reason = %q, want %q", refused.Reason, tt.reason)
			}
		})
	}
}
